crates/core/src/dom/ffi.rs

```diff
@@ -26,87 +29,90 @@ impl Document {
     #[uniffi::constructor]
     pub fn parse(input: String) -> Result<Arc<Self>, ParseError> {
         Ok(Arc::new(Self {
-            inner: Arc::new(SyncUnsafeCell::new(super::Document::parse(input)?)),
+            inner: Arc::new(Mutex::new(super::Document::parse(input)?)),
         }))
     }
 
     #[uniffi::constructor]
     pub fn empty() -> Arc<Self> {
         Arc::new(Self {
-            inner: Arc::new(SyncUnsafeCell::new(super::Document::empty())),
+            inner: Arc::new(Mutex::new(super::Document::empty())),
         })
     }
 
     #[uniffi::constructor]
     pub fn parse_fragment_json(input: String) -> Result<Arc<Self>, RenderError> {
-        let inner = Arc::new(SyncUnsafeCell::new(super::Document::parse_fragment_json(
-            input,
-        )?));
+        let inner = Arc::new(Mutex::new(super::Document::parse_fragment_json(input)?));
         Ok(Arc::new(Self { inner }))
     }
 
     pub fn set_event_handler(&self, handler: Box<dyn DocumentChangeHandler>) {
-        self.inner_mut().event_callback = Some(Arc::from(handler));
+        self.inner.lock().unwrap().event_callback = Some(Arc::from(handler));
     }
 
     pub fn merge_fragment_json(&self, json: &str) -> Result<(), RenderError> {
-        self.inner_mut().merge_fragment_json(json)
+        self.inner.lock().unwrap().merge_fragment_json(json)
     }
 
     pub fn root(&self) -> Arc<NodeRef> {
-        self.inner().root().into()
+        self.inner.lock().unwrap().root().into()
     }
 
     pub fn get_parent(&self, node_ref: Arc<NodeRef>) -> Option<Arc<NodeRef>> {
-        self.inner()
+        self.inner
+            .lock()
+            .unwrap()
             .parent(*node_ref)
             .map(|node_ref| node_ref.into())
     }
 
     pub fn children(&self, node_ref: Arc<NodeRef>) -> Vec<Arc<NodeRef>> {
-        self.inner()
+        self.inner
+            .lock()
+            .unwrap()
             .children(*node_ref)
             .iter()
             .map(|node| Arc::new(*node))
             .collect()
     }
 
     pub fn get_attributes(&self, node_ref: Arc<NodeRef>) -> Vec<Attribute> {
-        self.inner().attributes(*node_ref).to_vec()
+        self.inner.lock().unwrap().attributes(*node_ref).to_vec()
     }
+
     pub fn get(&self, node_ref: Arc<NodeRef>) -> NodeData {
-        self.inner().get(*node_ref).clone()
+        self.inner.lock().unwrap().get(*node_ref).clone()
```

Comment:
IMO use `expect("failed to get lock for document mutex")` rather than `unwrap()`. These calls always happen from the main thread and is single threaded so it shouldn't result in a deadlock but if it does a message will be more helpful than the panic and line number.